src/ppx_tools/dumpast.ml

```diff
@@ -0,0 +1,396 @@
+      Jsonoo.Encode.object_ (("type", Jsonoo.Encode.string label) :: args)
+
+    method other _ _ =
+      Jsonoo.Encode.string "serializing other values is not supported yet"
+
+    method nativeint _ _ =
+      Jsonoo.Encode.string "serializing nativeint is not supported yet"
+
+    method int64 _ _ =
+      Jsonoo.Encode.string "serializing int64 is not supported yet"
+
+    method int32 _ _ =
+      Jsonoo.Encode.string "serializing int32 is not supported yet"
+
+    method int value _ = Jsonoo.Encode.int value
+
+    method list f l l' =
+      (*FIXME different length?*)
+      if List.length l > List.length l' then
+        List.map2 f l l |> Jsonoo.Encode.list (fun x -> x)
+      else if List.length l < List.length l' then
+        List.map2 f l' l' |> Jsonoo.Encode.list (fun x -> x)
+      else
+        List.map2 f l l' |> Jsonoo.Encode.list (fun x -> x)
+
+    method option f o o' =
+      match (o, o') with
+      | Some x, Some x' ->
+        let a = f x x' in
+        a
+      | None, None ->
+        Jsonoo.Encode.object_ [ ("type", Jsonoo.Encode.string "None") ]
+      | _ -> raise (Traverse_ast2.Reparse_error "option")
+
+    method constr label args =
+      match args with
+      | [] -> Jsonoo.Encode.object_ [ ("type", Jsonoo.Encode.string label) ]
+      | _ ->
+        Jsonoo.Encode.object_ (("type", Jsonoo.Encode.string label) :: args)
+
+    method char value _ = Jsonoo.Encode.char value
+
+    method! location : location -> location -> Jsonoo.t =
+      fun { loc_start; loc_end; loc_ghost }
+          { loc_start = loc_start'; loc_end = loc_end'; loc_ghost = loc_ghost' } ->
+        let loc_start = super#position loc_start loc_start in
+        let real_loc_start = super#position loc_start' loc_start' in
+        let loc_end = super#position loc_end loc_end in
+        let real_loc_end = super#position loc_end' loc_end' in
+        let loc_ghost = self#bool loc_ghost loc_ghost' in
+        self#record "Location.t"
+          [ ("loc_start", real_loc_start)
+          ; ("origin_loc_start", loc_start)
+          ; ("loc_end", real_loc_end)
+          ; ("origin_loc_end", loc_end)
+          ; ("loc_ghost", loc_ghost)
+          ]
+
+    method! open_infos _a { popen_expr; popen_override; popen_loc; _ }
+        { popen_expr = popen_expr'
+        ; popen_override = popen_override'
+        ; popen_loc = popen_loc'
+        ; _
+        } =
+      let popen_expr = _a popen_expr popen_expr' in
+      let popen_override = self#override_flag popen_override popen_override' in
+      let popen_loc = self#location popen_loc popen_loc' in
+      let popen_attributes = Jsonoo.Encode.null in
+      (*FIXME*)
+      self#record "open_infos"
+        [ ("popen_expr", popen_expr)
+        ; ("popen_override", popen_override)
+        ; ("popen_loc", popen_loc)
+        ; ("popen_attributes", popen_attributes)
+        ]
+
+    method! directive_argument_desc x x' =
+      try super#directive_argument_desc x x' with
```

Comment:
Could you extract this try/catch pattern into a function:

```ocaml
val with_reparse : (unit -> unit) -> on_reparse_error:(unit -> unit)
```

Reply:
Yeah, pretty much. I pushed a commit that does this and it does clean things up a little.

Reply:
I don't think your commit went through for some reason. Could you please recheck?

Reply:
Oops. Done.